packages/logging/src/logging-and-telemetry.ts

```diff
@@ -82,30 +103,72 @@ class LoggingAndTelemetry implements MongoshLoggingAndTelemetry {
   private readonly mongoshVersion: string;
 
   private log: MongoLogWriter;
-  private pendingLogEvents: CallableFunction[] = [];
+  private pendingBusEvents: CallableFunction[] = [];
+  private pendingTelemetryEvents: CallableFunction[] = [];
   private isSetup = false;
-  private isBufferingEvents = false;
+  private isBufferingBusEvents = false;
+  private isBufferingTelemetryEvents = false;
+
+  private deviceId: string | undefined;
+  /** @internal */
+  public setupTelemetryPromise: Promise<void> = Promise.resolve();
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  private resolveDeviceId: (value: string) => void = () => {};
 
   constructor({
     bus,
     analytics,
     userTraits,
     mongoshVersion,
+    deviceId,
   }: MongoshLoggingAndTelemetryArguments) {
     this.bus = bus;
     this.analytics = analytics;
     this.log = LoggingAndTelemetry.dummyLogger;
     this.userTraits = userTraits;
     this.mongoshVersion = mongoshVersion;
+    this.deviceId = deviceId;
   }
 
   public setup(): void {
     if (this.isSetup) {
       throw new Error('Setup can only be called once.');
     }
+    this.isBufferingTelemetryEvents = true;
+
+    this.setupTelemetryPromise = this.setupTelemetry();
     this.setupBusEventListeners();
+
     this.isSetup = true;
-    this.isBufferingEvents = true;
+  }
+
+  public flush(): void {
+    // Run any telemetry events even if device ID hasn't been resolved yet
+    this.runAndClearPendingTelemetryEvents();
+
+    // Run any other pending events with the set or dummy log for telemetry purposes.
+    this.runAndClearPendingBusEvents();
+
+    this.resolveDeviceId('unknown');
+  }
+
+  private async setupTelemetry(): Promise<void> {
+    if (!this.deviceId) {
+      try {
+        this.deviceId ??= await Promise.race([
+          getDeviceId(),
+          new Promise<string>((resolve) => {
```

Comment:
This may not be necessary I am just not sure if there's an issue with leaving the `getDeviceId()` "running" while we're flushing the events; does this get killed? Don't want to end up delaying the shell exit because of this.

Reply:
I think it makes sense. Like you said probably not strictly necessary, but nice to have. Personally I'd leave a comment in this code with this reasoning for our future selves, but consider that a nit.